In [40]:
import numpy as np
import pandas as pd
#Matlab-style plotting
from matplotlib import pyplot as plt
%matplotlib inline
#Advance-style plotting
import seaborn as sns
color =sns.color_palette()
sns.set_style('darkgrid')
#Ignore annoying warning from sklearn and seaborn
import warnings
def ignore_warn(*args, **kwargs):
    pass
warnings.warn = ignore_warn

In [41]:
train = pd.read_excel(r'C:\Users\ABOFFICIAL\Desktop\Participants Data\Data_Train.xlsx', encoding='utf-8')
test = pd.read_excel(r'C:\Users\ABOFFICIAL\Desktop\Participants Data\Data_Test.xlsx', encoding='utf-8')

In [42]:
train.head(2)

Restaurant                             Location  \
0    ID_6321  FTI College, Law College Road, Pune   
1    ID_2882                 Sector 3, Marathalli   

                                 Cuisines Average_Cost Minimum_Order Rating  \
0  Fast Food, Rolls, Burger, Salad, Wraps         ₹200           ₹50    3.5   
1                     Ice Cream, Desserts         ₹100           ₹50    3.5   

  Votes Reviews Delivery_Time  
0    12       4    30 minutes  
1    11       4    30 minutes

In [43]:
print("{}" .format(train.shape))
print("{}" .format(test.shape))

(11094, 9)
(2774, 8)


There are 11094 observations and 9 columns in the train data including the target variable
There are 2774 observations and 8 columns in the test data excluding te target variable

In [44]:
train.dtypes

Restaurant       object
Location         object
Cuisines         object
Average_Cost     object
Minimum_Order    object
Rating           object
Votes            object
Reviews          object
Delivery_Time    object
dtype: object

In [45]:
test.dtypes

Restaurant       object
Location         object
Cuisines         object
Average_Cost     object
Minimum_Order    object
Rating           object
Votes            object
Reviews          object
dtype: object

Comment: All data types are object, This need to be preprocessed with the likes of Reviews and votes which are meant to be int but are objects

Before we could start exploratoring the data, it needs to be processed andd clean

In [46]:
train['Cuisines'].unique()

array(['Fast Food, Rolls, Burger, Salad, Wraps', 'Ice Cream, Desserts',
       'Italian, Street Food, Fast Food', ...,
       'North Indian, Mughlai, Kebab, Chinese',
       'Chinese, Seafood, Asian, North Indian',
       'Tibetan, Chinese, Continental, Momos'], dtype=object)

In [47]:
test['Cuisines'].unique()

array(['North Indian, Chinese, Assamese', 'Biryani, Kebab', 'Fast Food',
       'Mithai, North Indian, Chinese, Fast Food, South Indian',
       'Chinese, Fast Food', 'North Indian, Chinese',
       'South Indian, Fast Food', 'Kerala, South Indian, Chinese',
       'North Indian', 'Mithai', 'North Indian, Fast Food', 'Chinese',
       'North Indian, Seafood, Chinese, Mughlai', 'Street Food',
       'Fast Food, North Indian', 'Bakery, Desserts',
       'North Indian, Mughlai, Chinese', 'Cafe, Continental, Italian',
       'North Indian, Mughlai, Afghan', 'Asian', 'Cafe, Fast Food',
       'Chinese, Thai', 'South Indian',
       'North Indian, Italian, Fast Food', 'Cafe, Fast Food, Continental',
       'Desserts, Bakery', 'North Indian, Chinese, European',
       'Cafe, Continental, Italian, North Indian, Chinese, Momos',
       'Pizza, Fast Food', 'Desserts', 'Bakery',
       'North Indian, South Indian, Chinese',
       'Fast Food, Lebanese, Rolls',
       'South Indian, Fast Food, Chi

In [48]:
train['Average_Cost'].unique()

array(['₹200', '₹100', '₹150', '₹250', '₹650', '₹350', '₹800', '₹50',
       '₹400', '₹600', '₹300', '₹750', '₹450', '₹550', '₹1,000', '₹500',
       '₹900', '₹1,200', '₹950', '₹850', '₹700', '₹1,150', 'for',
       '₹1,100', '₹1,400', '₹2,050'], dtype=object)

Comment: 'for' is not a value, its best to place it with -1 or -999 or 0

In [49]:
test['Average_Cost'].unique()

array(['₹350', '₹100', '₹200', '₹150', '₹300', '₹50', '₹250', '₹500',
       '₹650', '₹400', '₹550', '₹450', '₹600', '₹750', '₹850', '₹1,000',
       '₹700', '₹800', '₹1,200'], dtype=object)

In [50]:
train['Minimum_Order'].unique()

array(['₹50', '₹99', '₹0', '₹200', '₹450', '₹350', '₹79', '₹400', '₹199',
       '₹500', '₹250', '₹150', '₹90', '₹299', '₹300', '₹240', '₹89',
       '₹59'], dtype=object)

In [51]:
test['Minimum_Order'].unique()

array(['₹50', '₹99', '₹500', '₹0', '₹200', '₹149', '₹199', '₹399', '₹89'],
      dtype=object)

In [52]:
test['Cuisines'].unique()

array(['North Indian, Chinese, Assamese', 'Biryani, Kebab', 'Fast Food',
       'Mithai, North Indian, Chinese, Fast Food, South Indian',
       'Chinese, Fast Food', 'North Indian, Chinese',
       'South Indian, Fast Food', 'Kerala, South Indian, Chinese',
       'North Indian', 'Mithai', 'North Indian, Fast Food', 'Chinese',
       'North Indian, Seafood, Chinese, Mughlai', 'Street Food',
       'Fast Food, North Indian', 'Bakery, Desserts',
       'North Indian, Mughlai, Chinese', 'Cafe, Continental, Italian',
       'North Indian, Mughlai, Afghan', 'Asian', 'Cafe, Fast Food',
       'Chinese, Thai', 'South Indian',
       'North Indian, Italian, Fast Food', 'Cafe, Fast Food, Continental',
       'Desserts, Bakery', 'North Indian, Chinese, European',
       'Cafe, Continental, Italian, North Indian, Chinese, Momos',
       'Pizza, Fast Food', 'Desserts', 'Bakery',
       'North Indian, South Indian, Chinese',
       'Fast Food, Lebanese, Rolls',
       'South Indian, Fast Food, Chi

In [53]:
train['Rating'].unique()

array(['3.5', '3.6', '3.7', '3.2', '3.8', '4.0', '3.9', '4.2', '-', '2.8',
       '3.0', '3.3', '3.1', '4.7', '3.4', '4.1', 'NEW', '2.9', '4.6',
       '4.3', '2.6', '4.5', '4.4', '4.8', '2.4', '2.7', '2.5',
       'Opening Soon', '2.2', '4.9', '2.3', '2.1', 'Temporarily Closed'],
      dtype=object)

In [54]:
test['Rating'].unique()

array(['4.2', 'NEW', '3.6', '2.9', '3.4', '3.3', '3.8', '3.2', '3.1',
       '4.1', '3.9', '-', '3.7', '4.0', '4.3', '3.0', '3.5', '2.6', '2.8',
       '4.5', '2.7', '4.4', '2.1', '2.4', '4.7', '4.6', '2.5', '4.8',
       '2.3', 'Opening Soon'], dtype=object)

Comment: 'Temporarily Closed', 'NEW', 'Opening Soon' , '-' are not values, is best to replace it with -1, -999, 0 or mean value

In [55]:
train['Votes'].unique()

array(['12', '11', '99', ..., '290', '698', '3700'], dtype=object)

In [56]:
test['Votes'].unique()

array(['361', '-', '36', '66', '38', '16', '72', '46', '5', '123', '9',
       '75', '134', '4', '80', '277', '288', '290', '30', '659', '145',
       '61', '8', '94', '119', '242', '977', '168', '607', '527', '27',
       '20', '67', '128', '31', '291', '152', '528', '150', '91', '169',
       '275', '41', '18', '19', '467', '1372', '81', '239', '52', '10',
       '53', '1319', '7', '1094', '189', '35', '1003', '39', '12', '51',
       '6', '99', '24', '14', '283', '174', '15', '287', '13', '57', '32',
       '271', '122', '44', '29', '729', '25', '393', '1485', '580', '96',
       '367', '22', '413', '85', '755', '106', '929', '33', '95', '463',
       '109', '77', '48', '371', '64', '335', '498', '400', '78', '544',
       '23', '108', '1666', '11', '954', '34', '84', '526', '156', '133',
       '141', '93', '103', '406', '286', '2314', '116', '87', '245',
       '644', '346', '45', '819', '351', '259', '163', '151', '255',
       '185', '47', '570', '248', '54', '798', '1489', '140

In [57]:
train['Reviews'].unique()

array(['4', '30', '95', '235', '18', '31', '1235', '26', '294', '61',
       '1239', '-', '128', '5', '53', '47', '46', '23', '160', '2', '1',
       '38', '9', '19', '7', '132', '74', '13', '111', '499', '29', '34',
       '851', '513', '22', '27', '32', '25', '37', '8', '35', '6', '173',
       '36', '15', '174', '11', '59', '16', '89', '170', '10', '1660',
       '71', '41', '20', '40', '3', '191', '105', '100', '106', '21',
       '65', '69', '2547', '159', '110', '42', '181', '109', '179', '490',
       '28', '343', '118', '690', '72', '296', '33', '164', '50', '129',
       '1581', '234', '45', '2363', '24', '115', '12', '138', '175', '39',
       '66', '1603', '1443', '390', '108', '1457', '112', '107', '259',
       '269', '52', '196', '794', '589', '48', '145', '536', '247', '17',
       '49', '245', '87', '83', '54', '453', '318', '99', '205', '165',
       '1845', '766', '114', '44', '51', '104', '478', '70', '232', '435',
       '14', '659', '554', '399', '77', '88', '503',

In [58]:
test['Reviews'].unique()

array(['225', '-', '16', '33', '14', '11', '10', '29', '2', '63', '1',
       '38', '47', '143', '147', '178', '4', '404', '12', '3', '26', '39',
       '51', '163', '627', '61', '102', '323', '242', '5', '37', '62',
       '19', '174', '20', '295', '64', '22', '153', '13', '7', '6', '228',
       '609', '40', '32', '27', '24', '659', '249', '59', '278', '17',
       '67', '93', '9', '155', '35', '15', '69', '45', '274', '752', '8',
       '378', '41', '217', '279', '444', '633', '66', '283', '21', '97',
       '48', '237', '271', '28', '195', '876', '575', '43', '170', '25',
       '200', '56', '53', '111', '77', '57', '120', '441', '165', '34',
       '42', '86', '18', '23', '319', '346', '664', '80', '30', '246',
       '204', '88', '232', '91', '49', '103', '60', '58', '90', '197',
       '71', '44', '114', '156', '1260', '241', '107', '127', '133', '70',
       '50', '179', '189', '468', '92', '503', '152', '1409', '2174',
       '104', '1392', '220', '72', '101', '175', '542', '3

In [59]:
train['Delivery_Time'].unique()

array(['30 minutes', '65 minutes', '45 minutes', '10 minutes',
       '20 minutes', '120 minutes', '80 minutes'], dtype=object)

since we have carfully look into the data, lets make some cleaning

In [60]:
train.head()

Restaurant                             Location  \
0    ID_6321  FTI College, Law College Road, Pune   
1    ID_2882                 Sector 3, Marathalli   
2    ID_1595                       Mumbai Central   
3    ID_5929                      Sector 1, Noida   
4    ID_6123   Rmz Centennial, I Gate, Whitefield   

                                 Cuisines Average_Cost Minimum_Order Rating  \
0  Fast Food, Rolls, Burger, Salad, Wraps         ₹200           ₹50    3.5   
1                     Ice Cream, Desserts         ₹100           ₹50    3.5   
2         Italian, Street Food, Fast Food         ₹150           ₹50    3.6   
3          Mughlai, North Indian, Chinese         ₹250           ₹99    3.7   
4                         Cafe, Beverages         ₹200           ₹99    3.2   

  Votes Reviews Delivery_Time  
0    12       4    30 minutes  
1    11       4    30 minutes  
2    99      30    65 minutes  
3   176      95    30 minutes  
4   521     235    65 minutes

In [61]:
train['Average_Cost'].apply(type)

0        <class 'str'>
1        <class 'str'>
2        <class 'str'>
3        <class 'str'>
4        <class 'str'>
5        <class 'str'>
6        <class 'str'>
7        <class 'str'>
8        <class 'str'>
9        <class 'str'>
10       <class 'str'>
11       <class 'str'>
12       <class 'str'>
13       <class 'str'>
14       <class 'str'>
15       <class 'str'>
16       <class 'str'>
17       <class 'str'>
18       <class 'str'>
19       <class 'str'>
20       <class 'str'>
21       <class 'str'>
22       <class 'str'>
23       <class 'str'>
24       <class 'str'>
25       <class 'str'>
26       <class 'str'>
27       <class 'str'>
28       <class 'str'>
29       <class 'str'>
             ...      
11064    <class 'str'>
11065    <class 'str'>
11066    <class 'str'>
11067    <class 'str'>
11068    <class 'str'>
11069    <class 'str'>
11070    <class 'str'>
11071    <class 'str'>
11072    <class 'str'>
11073    <class 'str'>
11074    <class 'str'>
11075    <class 'str'>
11076    <c

In [62]:
train['Average_Cost'] =train.Average_Cost.replace('[₹,]', '', regex =True)
test['Average_Cost'] =test.Average_Cost.replace('[₹,]', '', regex =True)

In [63]:
train['Minimum_Order'] =train.Minimum_Order.replace('[₹]', '', regex =True)
test['Minimum_Order'] =test.Minimum_Order.replace('[₹]', '', regex =True)

In [64]:
train['Rating'] =train.Rating.replace('[- NEW Opening Soon Temporarily Closed]', -1, regex =True)
test['Rating'] =test.Rating.replace('[- NEW Opening Soon Temporarily Closed]', -1, regex =True)

In [65]:
train['Delivery_Time'] =train.Delivery_Time.replace('[ minutes]', '', regex =True)

In [66]:
train['Votes'] =train.Votes.replace('[-]', -999, regex =True)
test['Votes'] =test.Votes.replace('[-]', -999, regex =True)

In [67]:
train['Reviews'] =train.Reviews.replace('[-]', -999, regex =True)
test['Reviews'] =test.Reviews.replace('[-]', -999, regex =True)

In [68]:
train['Reviews']=train['Reviews'].astype(int)
test['Reviews']=test['Reviews'].astype(int)

train['Votes']=train['Votes'].astype(int)
test['Votes']=test['Votes'].astype(int)

train['Rating']=train['Rating'].astype(float)
test['Rating']=test['Rating'].astype(float)

train['Minimum_Order']=train['Minimum_Order'].astype(int)
test['Minimum_Order']=test['Minimum_Order'].astype(int)

train['Delivery_Time']=train['Delivery_Time'].astype(int)

In [69]:
train.dtypes

Restaurant        object
Location          object
Cuisines          object
Average_Cost      object
Minimum_Order      int32
Rating           float64
Votes              int32
Reviews            int32
Delivery_Time      int32
dtype: object

In [70]:
train['Average_Cost'].unique()

array(['200', '100', '150', '250', '650', '350', '800', '50', '400',
       '600', '300', '750', '450', '550', '1000', '500', '900', '1200',
       '950', '850', '700', '1150', 'for', '1100', '1400', '2050'],
      dtype=object)

In [71]:
train[train['Average_Cost'] == 'for']

Restaurant         Location   Cuisines Average_Cost  Minimum_Order  \
6297    ID_6472  Pune University  Fast Food          for             50   

      Rating  Votes  Reviews  Delivery_Time  
6297    -1.0   -999     -999             30

In [72]:
train.drop(train.loc[train['Average_Cost']  == 'for'].index, inplace=True)

In [73]:
train.dtypes

Restaurant        object
Location          object
Cuisines          object
Average_Cost      object
Minimum_Order      int32
Rating           float64
Votes              int32
Reviews            int32
Delivery_Time      int32
dtype: object

In [74]:
train['Average_Cost']=train['Average_Cost'].astype(int)
test['Average_Cost']=test['Average_Cost'].astype(int)

In [75]:
train.dtypes

Restaurant        object
Location          object
Cuisines          object
Average_Cost       int32
Minimum_Order      int32
Rating           float64
Votes              int32
Reviews            int32
Delivery_Time      int32
dtype: object

In [76]:
test.dtypes

Restaurant        object
Location          object
Cuisines          object
Average_Cost       int32
Minimum_Order      int32
Rating           float64
Votes              int32
Reviews            int32
dtype: object

In [77]:
train.head()

Restaurant                             Location  \
0    ID_6321  FTI College, Law College Road, Pune   
1    ID_2882                 Sector 3, Marathalli   
2    ID_1595                       Mumbai Central   
3    ID_5929                      Sector 1, Noida   
4    ID_6123   Rmz Centennial, I Gate, Whitefield   

                                 Cuisines  Average_Cost  Minimum_Order  \
0  Fast Food, Rolls, Burger, Salad, Wraps           200             50   
1                     Ice Cream, Desserts           100             50   
2         Italian, Street Food, Fast Food           150             50   
3          Mughlai, North Indian, Chinese           250             99   
4                         Cafe, Beverages           200             99   

   Rating  Votes  Reviews  Delivery_Time  
0     3.5     12        4             30  
1     3.5     11        4             30  
2     3.6     99       30             65  
3     3.7    176       95             30  
4     3.2    521      235             65

In [78]:
train.to_excel(r'C:\Users\ABOFFICIAL\Desktop\Predicting-Food-Delivery-Time-master\Predicting-Food-Delivery-Time-master\processedtrain.xlsx',index=False)
test.to_excel(r'C:\Users\ABOFFICIAL\Desktop\Predicting-Food-Delivery-Time-master\Predicting-Food-Delivery-Time-master\processedtest.xlsx',index=False)